### ISSUE TITLE
List view scope does not apply in xslt list view webpart

### ISSUE THREAD

**Issue by digital88** [SubPointSolutions/spmeta2#1030]
### Brief description
Hello.
When I add list view to xsltListViewWebPart, web part does not have that view's scope. When I re-select same view in webpart settings in sharepoint page edit UI, scope is applied correctly.

### SharePoint API
* Sharepoint 2016 ENG + RUS lang pack, August 2016 CU

### SPMeta2 API
* SPMeta2 1.2.110 CSOM

### SPMeta2 model
```cs
var ct = new ContentTypeDefinition()
{
  Name = "Документ 2",
  Id = new Guid("guidString"),
  Group = "Grp",
  ParentContentTypeId = BuiltInContentTypeId.DocumentSet_Correct
};

var siteModel = SPMeta2Model.NewSiteModel(site =>
{
  site.AddContentType(ct);
});

var model = SPMeta2Model.NewWebModel(web =>
{
  var kbList = new ListDefinition()
  {
    Title = "KBase",
    TemplateType = BuiltInListTemplateTypeId.DocumentLibrary,
    CustomUrl = "Lists/Kbase",
    ContentTypesEnables = true,
    EnableVersioning = true,
  };

  var vw = new ListViewDefinition()
  {
    IsDefault = false,
    Title = "Последние",
    Url = "last10.aspx",
    Query = 
    "<OrderBy>" +
    $"<FieldRef Name=\"{BuiltInInternalFieldNames.Modified}\" Ascending=\"FALSE\" />" +
    "</OrderBy>",
    RowLimit = 10,
    Scope = BuiltInViewScope.Recursive,
    Fields = new Collection<string>()
    {
      BuiltInInternalFieldNames.LinkFilename,
      BuiltInInternalFieldNames.Modified,
      BuiltInInternalFieldNames._CheckinComment
    }
  };

  web.AddList(kbList, list =>
  {
    list.AddContentTypeLink(ct);
    list.AddListView(vw);
  });

  web.AddHostList(BuiltInListDefinitions.SitePages, list=>
  {
    list.AddWebPartPage(new WebPartPageDefinition()
    {
      Title = "Main",
      FileName = "Main.aspx",
      NeedOverride = true,
      CustomPageLayout = File.ReadAllText("layout.txt")
    }, home =>
    {
      home.AddWebPart(new XsltListViewWebPartDefinition()
      {
        ListUrl = "Lists/Kbase",
        ZoneId = "Body",
        ZoneIndex = 1,
        Title = "Last changes",
        TitleUrl = "~site/Lists/Kbase",
        Id = "g_+guid_string",
        ViewName = vw.Title,
        WebpartXmlTemplate = WebpartXmlExtensions.LoadWebpartXmlDocument(BuiltInWebPartTemplates.XsltListViewWebPart).SetListName("Kbase").ToString()
      });
    });
  };
});
```

**Comment by SubPointSupport** [comment]
Could be a CSOM limitation with list view bindings over ListView / XstlListView web pats. We'll investigate and let you know.

**Comment by andreasblueher** [comment]
I found the reason, but not yet the solution to this issue. 

`<property name="ViewFlags" type="Microsoft.SharePoint.SPViewFlags, Microsoft.SharePoint, Version=15.0.0.0, Culture=neutral, PublicKeyToken=71e9bce111e9429c">Html, TabularView, Hidden, Mobile</property>`

ViewFlags is not being set by SPMeta accordingly to the selected scope. However, even if you change the webpart xml to the needed value it's being ignored when being imported. Since the type is SPViewFlags I'm not sure if it's possible to set them via CSOM at all, haven't found any blog or code example doing so. However I found this [link](https://sharepoint.stackexchange.com/questions/67973/how-to-set-recurrencerowset-true-when-creating-a-view-programmatically) setting the property via SSOM (could have looked up SPMeta implementation too).

I've checked SSOM and it's working as expected there. Unless there is some magic way to define ViewFlags through client side code, I don't think it's possible. Not sure if you could warn people here, but for the actual listview it's working, so that's kind of an edge case.

**Comment by digital88** [comment]
How does this flags related to view scope? I can update view.Scope property via CSOM and have working code for setting xslt list view web-part scope by updating hidden view's scope in list views.

**Comment by andreasblueher** [comment]
View.Scope is one being set correctly, but this property has to be transmitted to the webpart aswell. SPMeta deploys the listview webpart by importing an xml using the webpart manager. Having a correct xml doesn't deploy the correct webpart using csom.

[Exports](https://gist.github.com/andreasblueher/620505cbe569b1b3d2139eaf6695e332)

Here you can see the xml which is generated by SPMeta (definition.txt) and the 2nd one is the export after manually editing the listviewwebpart (webpart.txt). They're almost the same, except for ViewFlags property.

You would need a way to edit an already placed listviewwebpart using CSOM.

**Comment by SubPointSupport** [comment]
> How are this flags related to view scope? I can update view.Scope property via CSOM and have working code for setting xslt list view web-part scope by updating hidden view's scope in list views.

@digital88, right, so you are working with the hidden view underneath the web part? Seems we need a post-fix for that property to make it transferred properly for CSOM.

**Comment by digital88** [comment]
Yes, like this
```cs
ClientContext ctx = ..;
var list = ctx.Web.List.GetByTitle("..");
ctx.Load(list, lst=>lst.Views);
ctx.ExecuteQuery();

var view = list.Views.Where(vw=>vw.Hidden && vw.Id == id).FirstOrDefault(); /* check id, title always equals string.Empty */

if (view != null)
{
  view.Scope = newScope;
  view.Update();
  ctx.ExecuteQuery();
}

**Comment by SubPointSupport** [comment]
Sweet, going to fix that, then. We'll update this ticket with the progress.

**Comment by andreasblueher** [comment]
list.Views provides GetById method so it might be more efficient to use that. 

[MSDN](https://msdn.microsoft.com/de-de/library/office/microsoft.sharepoint.client.viewcollection.getbyid.aspx)

Didn't think of the hidden views when I wrote my answer. I thought you would have to somehow change the webpart itself.

**Comment by SubPointSupport** [comment]
@andreasblueher, ListViewWebPart and XsltListViewWebPart create a hidden view on provisioning. For every such web part, there is a hidden list view in the target list created. 

All works well with SSOM, and only CSOM gives a lot of weird behavior with web part xml import so that tons of post-fixes on the hidden view is requred by the C# code.

**Comment by SubPointSupport** [comment]
Researching and testing your case, @digital88, thank you for reporting it. Below some comment on how it all works under the hood to get you a stable, predictable SharePoint experience.

SPMeta2 works very differently for CSOM implementation for both ListViewebPart and XsltListViewWebParts. By design, SharePoint CSOM API provides only web part XML import. In that case, XML gets imported some what magical way and only the default list view gets imported for these web parts. Again, only CSOM relates SharePoint issue.

We tried two implementations to get a workaround for this:
1) fixing list view after web part was provisioned, updating the list view and transferring properties from the original list view to newly created
2) making the target view default list view before web part import, make web part import, revert back the default view to the original one

After some testing, a year ago or even more, we decided to go with the second option - making the target view default one for the time being, import the web part and then revert back the default view. That way we were hoping to get the most "out of the box" experience by SharePoint without us getting into troubles of copying-pasting, transferring list view properties from the target one to the hidden one created for the web part. Seems, in some cases, even this trick does not work which is, frankly saying, extremely painful and disappointing. This is not the first time we put a workaround within SPMeta2 to make SharePoint work well, and it does not seem to be the last one.
